File: day5.py
#!/usr/bin/env python3
import re
from typing import Tuple, List, Dict

from util import load

LINE_REGEX = re.compile(r'^(\d+),(\d+) -> (\d+),(\d+)$')
Coord = Tuple[int, int]


class Line:
    _X = 0
    _Y = 1

    def __init__(self, start: Coord, end: Coord):
        self._start = start
        self._end = end

    def is_horizontal(self):
        return self._start[self._Y] == self._end[self._Y]

    def is_vertical(self):
        return self._start[self._X] == self._end[self._X]

    def is_horz_or_vert(self):
        return self.is_horizontal() or self.is_vertical()

    def points(self) -> List[Coord]:
        dx = self._end[self._X] - self._start[self._X]
        dy = self._end[self._Y] - self._start[self._Y]
        steps = max(abs(dx), abs(dy))
        coords: List[Coord] = []
        for i in range(steps):
            coords.append((self._start[self._X] + i * dx // steps,
                           self._start[self._Y] + i * dy // steps))
        coords.append(self._end)
        return coords


def map_line(line: str) -> Line:
    m = re.match(LINE_REGEX, line)
    if not m:
        raise Exception('invalid input: ' + line)
    return Line((int(m.group(1)), int(m.group(2))),
                (int(m.group(3)), int(m.group(4))))


def build_map(lines: List[Line], incl_diagonal=False) -> Dict[Coord, int]:
    line_map = {}
    for line in lines:
        if incl_diagonal or line.is_horz_or_vert():
            points = line.points()
            for point in points:
                count = line_map.get(point, 0)
                line_map[point] = count + 1
    return line_map


def count_overlaps(line_map: Dict[Coord, int]) -> int:
    return len(list(filter(lambda v: v > 1, line_map.values())))


if __name__ == "__main__":
    lines_data = load('day5.txt', mapper=map_line)

    map_data = build_map(lines_data)
    overlaps = count_overlaps(map_data)
    assert overlaps == 4873
    print("Horz/vert Overlaps = %d" % overlaps)

    map_data = build_map(lines_data, True)
    overlaps = count_overlaps(map_data)
    assert overlaps == 19472
    print("All overlaps = %d" % overlaps)

File: Elixir/apps/day5/lib/day5.ex
defmodule Day5 do
  @moduledoc """
  Documentation for `Day5`.
  """

  defmodule Coord do
    defstruct [:x, :y]
    @type t :: %__MODULE__{
                 x: integer,
                 y: integer,
               }

    @spec new(String.t()) :: Coord
    def new(line) do
      parsed = String.split(line, ",")
               |> Enum.map(&String.to_integer/1)
      %Coord {
        x: Enum.at(parsed, 0),
        y: Enum.at(parsed, 1)
      }
    end
  end

  defmodule Line do
    defstruct [:start, :end]
    @type t :: %__MODULE__{
                 start: Coord.t(),
                 end: Coord.t(),
               }

    @spec new(String.t()) :: Line
    def new(line) do
      coords = line
               |> String.split(" -> ")
               |> Enum.map(&Coord.new/1)
      %Line {
        start: Enum.at(coords, 0),
        end: Enum.at(coords, 1)
      }
    end

    def is_horizontal(line) do
      line.start.y == line.end.y
    end

    def is_vertical(line) do
      line.start.x == line.end.x
    end

    def is_horz_or_vert(line) do
      is_horizontal(line) or is_vertical(line)
    end

    @spec points(Line.t()) :: List.t(Coord.t())
    def points(line) do
      dx = line.end.x - line.start.x
      dy = line.end.y - line.start.y
      steps = max(abs(dx), abs(dy))
      0..steps
      |> Enum.map(
           fn i -> %Coord{
                     x: line.start.x + div(i * dx, steps),
                     y: line.start.y + div(i * dy, steps)
                   }
           end
         )
    end
  end

  @type field :: Map.t(Coord.t(), integer)
  @spec build_map(Line.t(), field) :: field
  def build_map(line, points) do
    Line.points(line)
    |> Enum.reduce(
         points,
         fn p, acc ->
           Map.update(acc, p, 1, &(&1 + 1))
         end
       )
  end

  def run(allow_diagonal) do
    File.read!("data/day5.txt")
    |> String.split("\n")
    |> Stream.filter(fn s -> String.length(s) != 0 end)
    |> Stream.map(&Line.new/1)
    |> Stream.filter(fn l -> allow_diagonal or Line.is_horz_or_vert(l) end)
    |> Enum.reduce(%{}, &build_map/2)
    |> Map.values()
    |> Stream.filter(&(&1 > 1))
    |> Enum.count()
  end
end
